pkg/clients/eks/eks.go

```diff
@@ -40,11 +40,15 @@ import (
 
 	"github.com/crossplane/provider-aws/apis/eks/v1beta1"
 	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+
+	smithyhttp "github.com/aws/smithy-go/transport/http"
 )
 
 const (
-	clusterIDHeader = "x-k8s-aws-id"
-	v1Prefix        = "k8s-aws-v1."
+	clusterIDHeader  = "x-k8s-aws-id"
+	expireHeader     = "X-Amz-Expires"
+	expireHeaderTime = "60"
```

Comment:
I don't know exactly how this connection info works. Does it create a kubeconfig with a static token? Is the token (and kubeconfig) only valid for 60s? Or is the token and kubeconfig rotated every 60s to make this work?

I think this PR is good but it would be cool if you could explain this :)

BTW how long did the token last before? 60s?

Reply:
It is possible to make requests to the Kubernetes API from a client that is outside the cluster, be that using the bare Kubernetes REST API or from one of the language specific Kubernetes clients (e.g., Go). 
In order to do so, we must create a bearer token that is included with the request to the API.
This bearer token requires you append the string `k8s-aws-v1.` with a base64 encoded string of a `signed HTTP request` to the `STS GetCallerIdentity Query API`. This is then sent it in the Authorization header of the request.
check:  https://github.com/kubernetes-sigs/aws-iam-authenticator#api-authorization-from-outside-a-cluster

==
kubeconfig in connection secret from eks in provider-aws - general overview:

```
kind: Secret
apiVersion: v1
data:
  clusterCA: <base64-cluster-ca>
  endpoint: <base64-endpoint>
  kubeconfig: <base64-kubeconfig>
```

**<= v0.19.0 (working authentication)**
we remove the base64 for the kubeconfig field and then we see a complete kubeconfig file - the interesting part now is the` token `under user - after string  `k8s-aws-v1.` we se the next base64 and i removed it

```
users:
- name: platform-ref-aws-cluster
  user:
    token: k8s-aws-v1.<base64>
```

then we see a presigned sts url with the following content:

```
https://sts.us-west-2.amazonaws.com/?
  Action=GetCallerIdentity&
  Version=2011-06-15&
  X-Amz-Algorithm=AWS4-HMAC-SHA256&
  X-Amz-Credential=xxxx%2F20211031%2Fus-west-2%2Fsts%2Faws4_request&
  X-Amz-Date=20211031T125725Z&
  X-Amz-Expires=60&
  X-Amz-SignedHeaders=host%3Bx-k8s-aws-id&
  X-Amz-Signature=67af338d83a46f2deeb2f2d15b58b23f21301595336ec613cc2b130b0d9a9d4e
```

**generate token via aws-cli**
i checked now how aws-cli generates the token 

```
aws eks get-token --cluster-name platform-ref-aws-cluster --region us-west-2
{"kind": "ExecCredential", "apiVersion": "client.authentication.k8s.io/v1alpha1", "spec": {}, "status": {"expirationTimestamp": "2021-11-01T19:16:17Z", "token": "k8s-aws-v1.<base64>"}}
```

so it looks like our token we using in the kubeconfig - so remove the base64:

```
https://sts.us-west-2.amazonaws.com/?
  Action=GetCallerIdentity&
  Version=2011-06-15&
  X-Amz-Algorithm=AWS4-HMAC-SHA256&
  X-Amz-Credential=xxxx%2F20211031%2Fus-west-2%2Fsts%2Faws4_request&
  X-Amz-Date=20211031T132212Z&
  X-Amz-Expires=60&
  X-Amz-SignedHeaders=host%3Bx-k8s-aws-id&
  X-Amz-Signature=67af338d83a46f2deeb2f2d15b58b23f21301595336ec613cc2b130b0daa87e4
```
looks like the same in our <= v0.19.0 aws-provider 


**>= v0.20.0**
we remove the base64 for the kubeconfig field and then we see a complete kubeconfig file - the interesting part now is the` token `under user - after string  `k8s-aws-v1.` we se the next base64 and i removed it

```
users:
- name: platform-ref-aws-cluster
  user:
    token: k8s-aws-v1.<base64>
```

```
https://sts.us-west-2.amazonaws.com/?
  Action=GetCallerIdentity&
  Version=2011-06-15&
  X-Amz-Algorithm=AWS4-HMAC-SHA256&
  X-Amz-Credential=xxxx%2F20211031%2Fus-west-2%2Fsts%2Faws4_request&
  X-Amz-Date=20211031T103835Z&
  X-Amz-SignedHeaders=host&
  X-Amz-Signature=ea57163ecb07f1154ac3d254b1aa2eb2247ad9d6aae5bc4ad3572d5246ad2d92
```

interesting in this version is that we see a _difference_ in `X-Amz-SignedHeaders` and a missing Header `X-Amz-Expires`

```
  X-Amz-Expires=60&
  X-Amz-SignedHeaders=host%3Bx-k8s-aws-id&
```
found one issue in aws sdkv2 github https://github.com/aws/aws-sdk-go-v2/issues/1137#issuecomment-789980978
and checked if we can add `x-k8s-aws-id`  again to `X-Amz-SignedHeaders` 
 in <= v0.19.0 we set also this header https://github.com/crossplane/provider-aws/blob/v0.19.0/pkg/clients/eks/eks.go#L328 

after that i could see in aws-authenticator log in cloudwatch:
`invalid X-Amz-Expires parameter in pre-signed URL: 0`

in <= v0.19.0 we set the X-Amz-Expires to 60
https://github.com/crossplane/provider-aws/blob/v0.19.0/pkg/clients/eks/eks.go#L337 (max time is 900 s)

that is the reason in this PR that i added both headers again:
https://github.com/crossplane/provider-aws/pull/901/files#diff-d0f093f49b3f8d568ade1abfbc36d6c3e7acfa65bc60977c63d5d9bbcfbbcb87R350-R351